src/core/ext/filters/client_channel/lb_policy/rls/rls.cc

```diff
@@ -1419,35 +1425,40 @@ void RlsLb::Cache::ResetAllBackoff() {
 void RlsLb::Cache::Shutdown() {
   map_.clear();
   lru_list_.clear();
-  grpc_timer_cancel(&cleanup_timer_);
+  if (cleanup_timer_handle_.has_value() &&
+      lb_policy_->channel_control_helper()->GetEventEngine()->Cancel(
+          *cleanup_timer_handle_)) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_lb_rls_trace)) {
+      gpr_log(GPR_INFO, "[rlslb %p] cache cleanup timer canceled", lb_policy_);
+    }
+    lb_policy_->Unref(DEBUG_LOCATION, "Shutdown");
+  }
+  cleanup_timer_handle_.reset();
 }
 
-void RlsLb::Cache::OnCleanupTimer(void* arg, grpc_error_handle error) {
-  Cache* cache = static_cast<Cache*>(arg);
-  cache->lb_policy_->work_serializer()->Run(
-      [cache, error]() {
-        RefCountedPtr<RlsLb> lb_policy(cache->lb_policy_);
+void RlsLb::Cache::OnCleanupTimer() {
+  lb_policy_->work_serializer()->Run(
```

Comment:
As above, please move the enterance into the `WorkSerializer` to be inline in the lambda (line 1374), so that we can just move the captured ref into it instead of doing this manual ref-tracking.

Reply:
This one is a behavioral change, but I think that's probably an improvement. The ref will now be taken earlier when scheduling a timer, rather than inside the work serializer callback. And if we don't explicitly release the ref before restarting the next timer, a ref is now held continuously between timer call execution and scheduling of the next timer, when previously the ref count would have dropped. I imagine if this is a significant change, we would have seen flakes in the previous implementation when trying to ref an orphaned lb_policy from inside the work serializer.

Reply:
I would prefer to use the same pattern we're using everywhere else of capturing the ref in the callback lambda, since that allows the compiler to enfore ref ownership.  I understand that it means that we'll take a new ref and then drop the previous one every time the timer fires, but this isn't performance-sensitive code, so that seems like a good trade-off.